// scholarly/dist/lib/googleScholar/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var axios_1 = __importDefault(require("axios"));
var search_1 = require("./search");
var user_1 = require("./user");
/**
 *
 *
 * @class _googleScholar
 * @extends {ArticleParser}
 */
var GoogleScholar = /** @class */ (function () {
    function GoogleScholar() {
        var _this = this;
        this.baseUrl = "https://scholar.google.com";
        this.searchTags = {
            results: ".gs_r",
            title: ".gs_ri h3 a",
            url: ".gs_ri h3 a",
            authors: ".gs_ri .gs_a",
            footers: ".gs_ri .gs_fl a",
            description: ".gs_ri .gs_rs",
            pdf: ".gs_ggsd a",
        };
        this.userTags = {
            results: ".gsc_a_tr",
            title: ".gsc_a_t a",
            url: ".gs_ri h3 a",
            authors: ".gs_gray",
            year: ".gsc_a_y",
            citations: ".gsc_a_c",
        };
        /**
         *
         *
         * @param {string} query
         */
        this.search = function (query) { return __awaiter(_this, void 0, void 0, function () {
            var articles, searchUrl, result, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        articles = [];
                        if (query === "") {
                            throw new Error("Query cannot be empty!");
                        }
                        searchUrl = encodeURI("/scholar?hl=en&q=".concat(query));
                        return [4 /*yield*/, axios_1.default.get(this.baseUrl + searchUrl)];
                    case 1:
                        result = _a.sent();
                        if (result.status !== 200) {
                            throw new Error(result.statusText);
                        }
                        data = result.data;
                        articles = (0, search_1.parseScholarArticle)(data, this.searchTags);
                        return [2 /*return*/, articles];
                }
            });
        }); };
        /**
         *
         *
         * @param {string} profile
         */
        this.user = function (profile) { return __awaiter(_this, void 0, void 0, function () {
            var articles, profileUrl, result, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        articles = [];
                        if (profile === "") {
                            throw new Error("User cannot be empty!");
                        }
                        profileUrl = encodeURI("/citations?hl=en&user=".concat(profile));
                        return [4 /*yield*/, axios_1.default.get(this.baseUrl + profileUrl)];
                    case 1:
                        result = _a.sent();
                        if (result.status !== 200) {
                            throw new Error(result.statusText);
                        }
                        data = result.data;
                        articles = (0, user_1.parseUserArticle)(data, this.userTags);
                        return [2 /*return*/, articles];
                }
            });
        }); };
    }
    return GoogleScholar;
}());
var googleScholar = new GoogleScholar();
exports.default = googleScholar;
